### PR TITLE
[SPARK-22384][SQL] Refine partition pruning when attribute is wrapped in Cast

### PR DESCRIPTION
**Pull request by jinxing64**
## What changes were proposed in this pull request?

Sql below will get all partitions from metastore, which put much burden on metastore;
```
CREATE TABLE `partition_test`(`col` int) PARTITIONED BY (`pt` byte)
SELECT * FROM partition_test WHERE CAST(pt AS INT)=1
```
The reason is that the the analyzed attribute `dt` is wrapped in `Cast` and `HiveShim` fails to generate a proper partition filter.
This pr proposes to take `Cast` into consideration when generate partition filter.

## How was this patch tested?
Test added.
This pr proposes to use analyzed expressions in `HiveClientSuite`

### CODE REVIEW

**Comment by jinxing64** [comment]
Could we fix this? Sql like above is common in my warehouse.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala [hunk 1]
```diff
@@ -53,7 +52,7 @@ class HiveClientSuite(version: String)
       for {
         ds <- 20170101 to 20170103
         h <- 0 to 23
-        chunk <- Seq("aa", "ab", "ba", "bb")
+        chunk <- Seq("11", "12", "21", "22")
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala]: What is the reason we need to change so many lines in the test case? Could you a new test case, instead of changing the existing one?

**Comment by gatorsmile** [comment]
For such cases, 
```Scala
CREATE TABLE test (value INT) PARTITIONED BY (dt STRING)
SELECT * from test where dt=2017
```

Can we just evaluate the right side `CAST(2017 as STRING)`, since it is foldable?

**Comment by jinxing64** [comment]
@gatorsmile 
Thanks a lot for your help :)

>Can we just evaluate the right side CAST(2017 as STRING), since it is foldable?

Do you mean to add a new rule ? -- cast the type of literal when partition key and the literal are different types.

In current code, `StringType` are promoted(e.g. casted to `IntegerType`) in `PromoteStrings` when it is BinaryComparison (https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L366)

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala (hunk 1 above)
Comment (jinxing64) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala]: Thanks a lot for review :)
I did two things:
1. Pass resolved `Expression` to generate filter -- because in current change, I need data type of the attribute;
2. "aa" -> "11", "ab" -> "12", "ba" -> "21", "bb" -> "22" -- because I need to pass an integer to set up the filter(chunk=11) and do the cast then.

**Comment by jinxing64** [comment]
@gatorsmile 
Thanks again for review this pr.

**Comment by gatorsmile** [comment]
The direction is right, but we need to restrict it to Cast (string_typed_attr as integral types) and the operator types must be either equal or not equal

**Comment by jinxing64** [comment]
True, I restrict it to be Cast (string_typed_attr as integral types) and `EqualTo`. `Not(EqualTo)` is not included, since the extra burden put to metastore is minor.

**Comment by jinxing64** [comment]
@gatorsmile
Thanks a lot for review this pr :)

**Comment by jinxing64** [comment]
I rebased this pr and resolved conflicts. 
cc @cloud-fan @jiangxb1987  Not sure if you have interest on this.
Take a look if have time. Thanks !

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 2]
```diff
@@ -657,18 +656,39 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
 
     val useAdvanced = SQLConf.get.advancedPartitionPredicatePushdownEnabled
 
+    object ExtractAttribute {
+      def unapply(expr: Expression): Option[Attribute] = {
+        expr match {
+          case attr: Attribute => Some(attr)
+          case cast @ Cast(child, dt: IntegralType, _) if child.dataType == StringType =>
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: can we generalize it? I think the hive analyzer can also do type coercion so any safe cast should be allowed, like int -> long, int -> string, etc.

We also need a config to safe guard it.

Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: seems we can use `Cast.mayTruncate`

**Comment by jinxing64** [comment]
@cloud-fan
Thanks a lot for looking into this.
I updated the change and generalized `ExtractAttribute`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala]: The first point looks fine, for the second one, can we generate new data for your new test case?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 3]
```diff
@@ -657,18 +656,46 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
 
     val useAdvanced = SQLConf.get.advancedPartitionPredicatePushdownEnabled
 
+    object ExtractAttribute {
+      def unapply(expr: Expression): Option[Attribute] = {
+        expr match {
+          case attr: Attribute => Some(attr)
+          case cast @ Cast(child, dt: StringType, _) if child.dataType.isInstanceOf[NumericType] =>
+            unapply(child)
+          case cast @ Cast(child, dt: NumericType, _) if child.dataType == StringType =>
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: I don't think this is safe. It assumes spark and hive has the same behavior when converting invalid string to numbers.

Comment (jinxing64) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: Are you worrying that Spark supports converting some invalid string to number but Hive doesn't? Yes, in that case, this change is dangerous. I didn't  think out an example, or there maybe such case in the future? Could you give an example?

On the other hand, it comes to me that this change is incorrect when there's precision truncation, e.g. cast("1.234" as int). But on a second thought, all the `ExtractAttribute` thing here happens within the scope of `getPartitionsByFilter` and such truncation would not happen?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala (hunk 1 above)
Comment (jinxing64) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala]: Thanks, I should refine the tests.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: I think we should only support safe upcast, `cast("1.234" as int)` should be excluded. Can we follow `Cast.mayTruncate` strictly?

**Comment by jinxing64** [comment]
@cloud-fan 
I removed some cases in `ExtractAttribute`. But to make it clear, do you mean that in below scenario, we don't prune partitions?
```
create table test(id int, string) partitioned by (dt string)
select count(1) from test where dt=20180101  -- HiveShim will get EqualTo(Cast(dt as int), 20180101) and Cast.mayTruncate will return true for this.
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 4]
```diff
@@ -657,18 +656,41 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
...
+      def unapply(expr: Expression): Option[Attribute] = {
+        expr match {
+          case attr: Attribute => Some(attr)
+          case cast @ Cast(child, dt, _) if !Cast.mayTruncate(child.dataType, dt) => unapply(child)
+          case _ => None
+        }
+      }
+    }
+
+    def decoratedValue(attr: Attribute, value: String): String = {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: what does it do? can you add some comments?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 5]
```diff
@@ -642,11 +641,11 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
           col.getType.startsWith(serdeConstants.CHAR_TYPE_NAME))
         .map(col => col.getName).toSet
 
-      def unapply(attr: Attribute): Option[String] = {
+      def unapply(attr: Attribute): Option[Attribute] = {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: why this change?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala (hunk 4 above)
Comment (jinxing64) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: e.g. 
dt="20180101"  if dt is of NumericType, we remove the quote and it changes to be  dt=20180101
dt=201880101 if dt is of StringType, we add the quote and it changes to be dt="20180101"

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala (hunk 5 above)
Comment (jinxing64) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: Because I need a Attribute to be passed into `decoratedValue`, which will process the Literal -- add or remove quotes if needed.

**Comment by cloud-fan** [comment]
> select count(1) from test where dt=20180101 

This should be forbidden, users should rewrite it into `dt="20180101"`

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 6]
```diff
@@ -657,18 +656,30 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
 
     val useAdvanced = SQLConf.get.advancedPartitionPredicatePushdownEnabled
 
+    object ExtractAttribute {
+      def unapply(expr: Expression): Option[Attribute] = {
+        expr match {
+          case attr: Attribute => Some(attr)
+          case cast @ Cast(child, dt, _) if !Cast.mayTruncate(child.dataType, dt) => unapply(child)
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: nit: `cast @ ` is not needed here.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala [hunk 7]
```diff
@@ -66,6 +65,19 @@ class HiveClientSuite(version: String)
     client
   }
 
+  private def attr(name: String): Attribute = {
+    name match {
+      case "ds" =>
+        AttributeReference("ds", IntegerType)()
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala]: This is error-prone, we should look up the attribute by name from `spark.table("test")`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala [hunk 8]
```diff
@@ -74,23 +86,23 @@ class HiveClientSuite(version: String)
   test(s"getPartitionsByFilter returns all partitions when $tryDirectSqlKey=false") {
     val client = init(false)
     val filteredPartitions = client.getPartitionsByFilter(client.getTable("default", "test"),
-      Seq(parseExpression("ds=20170101")))
+      Seq(EqualTo(attr("ds"), Literal(20170101, IntegerType))))
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala]: we can import `import org.apache.spark.sql.catalyst.dsl.expressions._` to simplify expression creation

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala [hunk 9]
```diff
@@ -207,65 +271,68 @@ class HiveClientSuite(version: String)
...
+    val expectedPartitions = expectedPartitionCubes.map(getPartitionsFromCube(_)).reduce(_ ++ _)
 
     val actualFilteredPartitionCount = filteredPartitions.size
 
     assert(actualFilteredPartitionCount == expectedPartitionCount,
       s"Expected $expectedPartitionCount partitions but got $actualFilteredPartitionCount")
-    assert(filteredPartitions.map(_.spec.toSet).toSet == expectedPartitions.toSet)
+    assert(filteredPartitions.map(_.spec).toSet == expectedPartitions.toSet)
+  }
+
```
Comment (jinxing64) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala]: Parse the cube to be a list of partitions, the result is like `Seq[Map[partition name, partition value]]`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala [hunk 10]
```diff
@@ -207,65 +271,68 @@ class HiveClientSuite(version: String)
...
+      expectedPartitionCubes: Seq[Map[String, Seq[Any]]]): Unit = {
+    testMetastorePartitionFiltering(table, filterExpr, expectedPartitionCubes, identity)
   }
 
   private def testMetastorePartitionFiltering(
-      filterString: String,
-      expectedPartitionCubes: Seq[(Seq[Int], Seq[Int], Seq[String])],
+      table: String,
+      filterExpr: Expression,
+      expectedPartitionCubes: Seq[Map[String, Seq[Any]]],
```
Comment (jinxing64) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala]: With this change, number of partition names in `expectedPartitionCubes`  is not necessarily to be 3. And schema of `expectedPartitionCubes` is like Seq[Map[partition name, partition values]]

**Comment by jinxing64** [comment]
@cloud-fan 
Sorry for late reply, so busy these days.
In current change:
1. I follow `Cast.mayTruncate` strictly when extract partition Attribute;
2. I created new test data in `HiveClientSuite` and `testMetastorePartitionFiltering` can be used for validation for tables with different partitions schema.

If needed, I can create more tests -- different binary comparisons and different datatypes.

**Comment by jinxing64** [comment]
And also I think we have same problem for datasource table.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala [hunk 11]
```diff
@@ -59,38 +61,62 @@ class HiveClientSuite(version: String)
...
     super.beforeAll()
     client = init(true)
   }
 
   test(s"getPartitionsByFilter returns all partitions when $tryDirectSqlKey=false") {
     val client = init(false)
-    val filteredPartitions = client.getPartitionsByFilter(client.getTable("default", "test"),
-      Seq(parseExpression("ds=20170101")))
+    val filteredPartitions = client.getPartitionsByFilter(client.getTable("default", "test0"),
+      Seq(EqualTo(pAttr("test0", "ds"), Literal(20170101, IntegerType))))
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala]: we can import `org.apache.spark.sql.catalyst.dsl.expressions._` to simplify expression creation

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala [hunk 12]
```diff
@@ -207,65 +271,68 @@ class HiveClientSuite(version: String)
...
     assert(actualFilteredPartitionCount == expectedPartitionCount,
       s"Expected $expectedPartitionCount partitions but got $actualFilteredPartitionCount")
-    assert(filteredPartitions.map(_.spec.toSet).toSet == expectedPartitions.toSet)
+    assert(filteredPartitions.map(_.spec).toSet == expectedPartitions.toSet)
+  }
+
+  private def getPartitionsFromCube(cube: Map[String, Seq[Any]]): Seq[Map[String, String]] = {
+    cube.map {
+      case (k: String, pts: Seq[Any]) => pts.map(pt => (k, pt.toString))
+    }.foldLeft(Seq(Seq[(String, String)]()))((seq0, seq1) => {
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala]: this is hard to read, please use loop and mutable states directly.

**Comment by cloud-fan** [comment]
In general cast is hard to be pushed into data source, e.g. `cast(a as string) = string` if a is int, how should data source handle it?

In the meanwhile, I think we can omit most of the cast in the format of `attribute = literal`. e.g. `cast(byteCol as int) = 0`, we know `0` is within byte range, we can convert it to `byteCol = (byte) 0`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala (hunk 11 above)
Comment (jinxing64) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala]: Thanks, with `org.apache.spark.sql.catalyst.dsl.expressions._`, code can be much cleaner.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala (hunk 12 above)
Comment (jinxing64) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala]: Hmm, it's  a recursion problem. I tried to use loop&mutable states directly, but seems not become easier readable;

In current change, I extract a `PartitionSpec` type and add comment. I think it's better now?

**Comment by cloud-fan** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 13]
```diff
@@ -657,17 +656,29 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
...
+        expr match {
+          case attr: Attribute => Some(attr)
+          case Cast(child, dt, _) if !Cast.mayTruncate(child.dataType, dt) => unapply(child)
+          case _ => None
+        }
+      }
+    }
+
     def convert(expr: Expression): Option[String] = expr match {
       case In(NonVarcharAttribute(name), ExtractableLiterals(values)) if useAdvanced =>
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: shall we apply the cast handling here?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientVersions.scala [hunk 14]
```diff
@@ -23,5 +23,5 @@ import org.apache.spark.SparkFunSuite
 
 private[client] trait HiveClientVersions {
   protected val versions =
-    IndexedSeq("0.12", "0.13", "0.14", "1.0", "1.1", "1.2", "2.0", "2.1", "2.2", "2.3")
+    IndexedSeq("1.2")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientVersions.scala]: why change this?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala [hunk 15]
```diff
@@ -46,10 +48,11 @@ class HiveClientSuite(version: String)
     val hadoopConf = new Configuration()
     hadoopConf.setBoolean(tryDirectSqlKey, tryDirectSql)
     val client = buildClient(hadoopConf)
-    client
-      .runSqlHive("CREATE TABLE test (value INT) PARTITIONED BY (ds INT, h INT, chunk STRING)")
+    client.runSqlHive("CREATE TABLE test0 (value INT) PARTITIONED BY (ds INT, h INT, chunk STRING)")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HiveClientSuite.scala]: can we reuse the existing data set to test? e.g. `'ds.cast(LongType) == 8L`. Then we can reduce the code diff here.

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by jinxing64** [comment]
Thanks for merging !
